File: leetcode/C_04_stack/D2024_01_12_do_2_Q735_asteroid_collision/Solution.java
package leetcode.C_04_stack.D2024_01_12_do_2_Q735_asteroid_collision;

import java.util.ArrayDeque;
import java.util.Deque;

public class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        int n = asteroids.length;
        Deque<Integer> stack = new ArrayDeque<>();

        for (int i = 0; i < n; i++) {
            int itemAsteroid = asteroids[i];
            if (itemAsteroid > 0) {
                stack.push(itemAsteroid);
            } else {
                // calculate negative sign case until the first item is negative or all the rest
                // are possitive
                while (!stack.isEmpty() && stack.peek() > 0 && stack.peek() + itemAsteroid < 0) {
                    stack.pop();
                }

                if (stack.isEmpty()) {
                    stack.push(itemAsteroid);
                } else if (stack.peek() + itemAsteroid == 0) {
                    stack.pop();
                } else if (stack.peek() < 0) {
                    stack.push(itemAsteroid);
                }
            }
        }

        int resSize = stack.size();
        int[] res = new int[resSize];
        for (int i = resSize - 1; i >= 0; i--) {
            res[i] = stack.pop();
        }

        return res;
    }
}

File: leetcode/C_04_stack/D2024_01_12_do_2_Q735_asteroid_collision/solution.go
package main

import "fmt"

func asteroidCollision(asteroids []int) []int {
	stack := []int{}
	n := len(asteroids)

	for i := 0; i < n; i++ {
		itemAsteroid := asteroids[i]
		if itemAsteroid > 0 { // positive direction
			stack = append(stack, itemAsteroid)
		} else {
			// calculate negative sign case until the first item is negative or all the rest are possitive
			for len(stack) != 0 && stack[len(stack)-1] > 0 && stack[len(stack)-1]+itemAsteroid < 0 {
				stack = stack[:len(stack)-1]
			}

			if len(stack) == 0 {
				stack = append(stack, itemAsteroid)
			} else if stack[len(stack)-1]+itemAsteroid == 0 {
				stack = stack[:len(stack)-1]
			} else if stack[len(stack)-1] < 0 {
				stack = append(stack, itemAsteroid)
			}
		}
	}

	resultSize := len(stack)
	result := make([]int, resultSize)
	for i := resultSize - 1; i >= 0; i-- {
		result[i] = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
	}

	return result
}

func main() {
	input := []int{}
	asteroidCollision(input)
	fmt.Println("run main")
}
